x-pack/plugin/analytics/src/test/java/org/elasticsearch/xpack/analytics/rate/InternalResetTrackingRateTests.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.analytics.rate;
+
+import org.elasticsearch.common.util.CollectionUtils;
+import org.elasticsearch.plugins.SearchPlugin;
+import org.elasticsearch.search.aggregations.Aggregation;
+import org.elasticsearch.search.aggregations.InternalAggregation;
+import org.elasticsearch.search.aggregations.ParsedAggregation;
+import org.elasticsearch.test.InternalAggregationTestCase;
+import org.elasticsearch.xcontent.NamedXContentRegistry;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xpack.analytics.AnalyticsPlugin;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.CoreMatchers.instanceOf;
+import static org.mockito.Mockito.mock;
+
+public class InternalResetTrackingRateTests extends InternalAggregationTestCase<InternalResetTrackingRate> {
+
+    @Override
+    protected SearchPlugin registerPlugin() {
+        return new AnalyticsPlugin();
+    }
+
+    @Override
+    protected InternalResetTrackingRate createTestInstance(String name, Map<String, Object> metadata) {
+        return new InternalResetTrackingRate(name, null, metadata, 0, 0, 0, 0, 0);
+    }
+
+    private static InternalResetTrackingRate rate(double startValue, double endValue, long startTime, long endTime, double resetComp) {
+        return new InternalResetTrackingRate("n", null, null, startValue, endValue, startTime, endTime, resetComp);
+    }
+
+    public void testReduction() {
+        List<InternalAggregation> rates = List.of(
+            rate(0, 10, 1000, 2000, 0),
+            rate(10, 20, 2000, 3000, 0),
+            rate(20, 5, 3000, 4000, 25), // internal reset
+            rate(5, 15, 4000, 5000, 0),
+            rate(0, 10, 5000, 6000, 0)  // cross-boundary reset
+        );
+        InternalAggregation reduced = rates.get(0).reduce(rates, null);
+        assertThat(reduced, instanceOf(Rate.class));
+        assertThat(((Rate) reduced).getValue(), equalTo(0.01));
+    }
+
+    @Override
+    protected void assertReduced(InternalResetTrackingRate reduced, List<InternalResetTrackingRate> inputs) {
+        for (InternalResetTrackingRate input : inputs) {
+            assertEquals(0.01f, input.getValue(), 0.001);
+        }
+        assertEquals(0.01f, reduced.getValue(), 0.001);
+    }
+
+    @Override
+    protected boolean supportsOutOfOrderReduce() {
+        return false;
+    }
```

Comment:
So in order to detect resets between `InternalTrackingRate` instances, these instances will need to be merged in order. I think most of the time no reduce will happen for `rate` agg. This because it needs to be wrapped in a `time_series` agg. Only for time series agg buckets with same key (same tsid) will need to do a reduce. So when the range of the parent bucket of time series bucket crosses backing index boundary then resets may not be detected.

But in any case the `InternalResetTrackingRate` instance need to be merged in order and with partial reduce I think we can't guarantee that? Maybe this `supportsOutOfOrderReduce` should be a flag in `InternalAggregation`? This can then be set by `InternalResetTrackingRate` and then no partial reduce will be triggered?

Reply:
This agg will work with partial reduce *as long as the reduction is in-order* - in other words, we can reduce buckets from time 1 - 10, and then further reduce that combined bucket with buckets 11 - 20, but we can't reduce buckets 1, 2, 4, 7 and 9, and then try and merge in bucket 3.  

As I understand it, partial reduction is used by CCS and by async search; async search will always reduce in order so that it can provide meaningful partial results, so we have no problem there.  And TSDB doesn't support tsids across clusters, so the CCS situation should also not apply here.

The `supportsOutOfOrderReduce` flag doesn't stop us testing partial reductions, it stops us selecting a random set of internal buckets to reduce.  When this flag is set to `false`, partial reductions will always be in-order.

Reply:
> partial reduction is used by CCS and by async search; async search will always reduce in order so that it can provide meaningful partial results, so we have no problem there. 

👍 

(I think partial reduce also happens during regular search depending `SearchRequest#batchedReduceSize`)

> And TSDB doesn't support tsids across clusters, so the CCS situation should also not apply here.

True, a tsid across data streams or clusters is impossible.

> The supportsOutOfOrderReduce flag doesn't stop us testing partial reductions, it stops us selecting a random set of internal buckets to reduce. When this flag is set to false, partial reductions will always be in-order.

Agreed, then my previous comment about having a flag `InternalAggregation` makes no sense and can be ignored.

Reply:
I hadn't realized that async search would reduce in order. Ok.